### DISCUSSION TITLE
Body & Query for PATCH/PUT methods

### DISCUSSION THREAD

**Discussion by shubham941122** [RobinTail/express-zod-api#2097] [Q&A] [answered]
Hi, is there a provision to take query, path, headers and requestBody as input for these methods.

```ts
inputSources: {
    get: ['query', 'params', 'headers'],
    post: ['body', 'params', 'query', 'headers'],
    put: ['body', 'query', 'params', 'headers'],
    patch: ['body', 'query', 'params', 'headers'],
    delete: ['query', 'params', 'headers'],
  },
```

Endpoint:
```ts
export const updateUserEndpoint = taggedEndpointsFactory.addMiddleware(methodProviderMiddleware).build({
  method: 'post',
  tag: 'users',
  shortDescription: 'Updates an user by its ID.',
  description: 'Example user retrieval endpoint.',
  input: z
    .object({
      ...headersSchema.shape,
      id: z
        .string()
        .trim()
        .regex(/\d+/)
        .transform((id) => parseInt(id, 10))
        .describe('a numeric string containing the id of the user'),
      name: z.string().min(1),
      serviceInfo: z.object({
        serviceBandwidth: z.string().trim().describe('a string containing the serviceBandwidth'),
        serviceType: z.string().trim().describe('a string containing the serviceType'),
      }),
    })
    .example({
      id: '12',
      name: 'John Doe',
      serviceInfo: {
        serviceBandwidth: '10GB',
        serviceType: 'Fiber',
      },
    }),
  output: z
    .object({
      name: z.string(),
      createdAt: ez.dateOut(),
    })
    .example({
      name: 'John Doe',
      createdAt: new Date('2021-12-31'),
    }),
  handler: async ({ input: { id, name }, options: { method }, logger }) => {
    logger.debug(`Requested id: ${id}, method ${method}, Name to set: ${name}`);

    if (id > 100) throw createHttpError(404, 'User not found');

    throw createHttpError(500, 'Not implemented yet!');
  },
});
```

With above config, i am not able to achieve the desired behaviour.  Any suggestions?

**Comment by RobinTail** [comment, marked as answer]
> Hi, is there a provision to take query, path, headers and requestBody as input for these methods.

- query ✅ 
- path ❌ (make a middleware providing it within `options`), but path params ✅  — use `params`
- headers ✅*
  - custom ones only, see: https://github.com/RobinTail/express-zod-api/blob/master/README.md#headers-as-input-source
- requestBody — is just `body` ✅ 
  - https://github.com/RobinTail/express-zod-api/blob/master/README.md#customizing-input-sources

> With above config, i am not able to achieve the desired behaviour. Any suggestions?

- What is the desired behavior exactly?
- What is happening instead, @shubham941122 ?

**Reply by shubham941122** [reply]
I want to accept all input sources in these HTTP Methods. 
But i am not able to accept these all at once.

I want to accept  as below :
1. name & serviceInfo as requestBody.
2. key as query param
3. id as pathParam
4. Headers working as expected.

![image](https://github.com/user-attachments/assets/004d4df4-55e9-453a-a58a-7ee3c183a00f)

**Reply by RobinTail** [reply]
@shubham941122 , you define the [properties](https://expressjs.com/en/5x/api.html#req) of `Request` in the config property `inputSources` per method.
The data coming from those properties is combined/merged together into a single object being validated against your `input` schema.

All the required properties you need are available: `body`, `query`, `params` and `headers`, so that it should work, unless something else is interfering, or your schema does not match the request you're sending, or you're sending it in a some wrong way, or you're checking not the thing you're sending, IDK 

> But i am not able to accept these all at once.

By doing what exactly you came to that conclusion?
How exactly you decide on that ability? How do you check it?
Without a complete reproducible sample including CURL request I can't figure out what's happening on your side.

**Reply by shubham941122** [reply]
Thanks @RobinTail for the explanation. Let me get back to you if any further queries.
